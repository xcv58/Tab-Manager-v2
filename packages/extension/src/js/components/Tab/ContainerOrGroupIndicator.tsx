import { observer } from 'mobx-react-lite'

let ContainerOrGroupIndicator = (): React.ReactNode => null

if (process.env.TARGET_BROWSER === 'firefox') {
  ContainerOrGroupIndicator = observer(require('./_ContainerIndicator').default)
}

if (process.env.TARGET_BROWSER === 'chrome') {
  ContainerOrGroupIndicator = observer(require('./_TabGroupIndicator').default)
}

export default ContainerOrGroupIndicator
