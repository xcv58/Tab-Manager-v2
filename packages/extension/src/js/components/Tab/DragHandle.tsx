import React from 'react'
import classNames from 'classnames'
import { useTheme } from 'components/hooks/useTheme'

export default () => {
  const isDarkTheme = useTheme()
  return (
    <button
      className={classNames(
        'inline-flex items-center justify-center w-8 h-8 m-2 rounded-full hover:shadow-xs focus:outline-none focus:ring cursor-move',
        {
          'hover:bg-blue-200 active:bg-blue-300': !isDarkTheme,
          'hover:bg-gray-600 active:bg-gray-800': isDarkTheme,
        }
      )}
    >
      &#9776;
    </button>
  )
}
