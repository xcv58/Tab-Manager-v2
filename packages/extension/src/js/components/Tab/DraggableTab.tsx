import React, { useEffect } from 'react'
import classNames from 'classnames'
import { observer } from 'mobx-react-lite'
import { useDrag } from 'react-dnd'
import DroppableTab from './DroppableTab'
import { ItemTypes } from 'libs/react-dnd'
import { useStore } from 'components/hooks/useStore'
import { getEmptyImage } from 'react-dnd-html5-backend'
import { TabProps } from '../types'

export default observer((props: TabProps) => {
  const { tab } = props
  const { dragStore } = useStore()
  const [dragProps, drag, connectDragPreview] = useDrag({
    item: { type: ItemTypes.TAB },
    canDrag: true,
    begin: () => {
      dragStore.dragStart(tab)
    },
    end: () => {
      dragStore.dragEnd()
    },
    isDragging: () => tab.isSelected,
    collect: (monitor) => {
      return {
        isDragging: monitor.isDragging(),
      }
    },
  })
  const { isDragging } = dragProps
  useEffect(() => {
    connectDragPreview(getEmptyImage())
  })
  return (
    <div ref={drag} className={classNames({ 'opacity-25': isDragging })}>
      <DroppableTab {...props} />
    </div>
  )
})
