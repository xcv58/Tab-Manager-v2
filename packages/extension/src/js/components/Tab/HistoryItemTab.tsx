import React, { useCallback } from 'react'
import { observer } from 'mobx-react-lite'
import { HistoryItem } from 'stores/SearchStore'
import { match } from 'fuzzy'
import classNames from 'classnames'
import Url from 'components/Tab/Url'
import { useStore } from 'components/hooks/useStore'
import moment from 'moment'
import { openURL } from 'libs'

const pre = "<span class='text-red-500'>"
const post = '</span>'

type Props = { tab: HistoryItem }

const Content = observer((props: Props) => {
  const { query } = useStore().searchStore
  const { title } = props.tab
  const getHighlightNode = useCallback(
    (text) => {
      if (!query) {
        return text
      }
      const result = match(query, text, { pre, post })
      if (!result) {
        return <div>{text}</div>
      }
      return <div dangerouslySetInnerHTML={{ __html: result.rendered }} />
    },
    [query]
  )
  const buttonClassName = classNames(
    'group flex flex-col justify-center flex-1 h-12 overflow-hidden text-left m-0 rounded-sm text-base'
  )
  return (
    <button className={buttonClassName} disabled>
      <div className="w-full overflow-hidden truncate">
        {getHighlightNode(title)}
      </div>
      <Url {...props} {...{ getHighlightNode }} />
    </button>
  )
})

export default observer(function HistoryItemTab(props: Props) {
  const { tab } = props
  const { lastVisitTime, typedCount, visitCount } = props.tab

  return (
    <div
      tabIndex={-1}
      className="relative flex items-center w-full pl-4 group"
      onClick={() => openURL(tab.url)}
    >
      <Content tab={tab} />
      <div className="flex-col px-2 text-sm opacity-75 group-hover:opacity-100">
        <div>{moment(lastVisitTime).fromNow()}</div>
        <div className="text-right">
          {typedCount}/{visitCount}
        </div>
      </div>
    </div>
  )
})
