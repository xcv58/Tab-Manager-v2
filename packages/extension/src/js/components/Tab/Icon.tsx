import React from 'react'
import classNames from 'classnames'
import { observer } from 'mobx-react-lite'
import Checkbox from '@material-ui/core/Checkbox'
import IconButton from '@material-ui/core/IconButton'
import { useStore } from 'components/hooks/useStore'
import { TabProps } from 'components/types'

const ARIA_LABEL = 'Toggle select'

export const Icon = observer((props: TabProps) => {
  const { userStore } = useStore()
  const { focus, select, iconUrl, isSelected, bulkSelect } = props.tab
  const checkbox = (
    <Checkbox
      color="primary"
      checked={isSelected}
      inputProps={{
        'aria-label': ARIA_LABEL,
      }}
      onClick={(e) => {
        if (process.env.TARGET_BROWSER === 'firefox') {
          if (e.altKey) {
            return props.tab.selectTabsInSameContainer()
          }
        }
        if (isSelected || !e.shiftKey) {
          select()
        } else {
          bulkSelect()
        }
      }}
    />
  )
  if (!userStore.showTabIcon) {
    return checkbox
  }
  return (
    <div className="group">
      <div
        className={classNames({
          hidden: isSelected,
          'group-hover:hidden': !isSelected,
        })}
      >
        <IconButton
          aria-label={ARIA_LABEL}
          className="focus:outline-none focus:ring"
          onClick={select}
          onFocus={focus}
        >
          <img className="w-6 h-6" src={iconUrl} />
        </IconButton>
      </div>
      <div
        className={classNames('focus:outline-none focus:ring', {
          'hidden group-hover:block': !isSelected,
        })}
      >
        {checkbox}
      </div>
    </div>
  )
})

export default observer((props: TabProps) => {
  const { faked, tab } = props
  const { iconUrl } = tab
  if (!faked) {
    return <Icon {...props} />
  }
  return (
    <div>
      <IconButton
        disabled
        aria-label={ARIA_LABEL}
        className="focus:outline-none focus:ring"
      >
        <img className="w-6 h-6" src={iconUrl} />
      </IconButton>
    </div>
  )
})
