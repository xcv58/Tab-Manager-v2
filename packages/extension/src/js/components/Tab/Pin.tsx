import React from 'react'

const PIN = (
  <div className="z-auto w-0 h-0 text-xs origin-bottom-right transform -rotate-90 translate-y-4 pointer-events-none">
    ðŸ“Œ
  </div>
)

export default PIN
