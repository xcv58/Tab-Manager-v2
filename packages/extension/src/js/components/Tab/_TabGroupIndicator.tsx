import React from 'react'
import { useStore } from 'components/hooks/useStore'
import Tab from 'stores/Tab'
import { TabGroup } from 'stores/TabGroupStore'

const _TabGroupIndicator = (props: Tab) => {
  const { groupId } = props
  const { tabGroupStore } = useStore()
  const tabGroup: TabGroup = tabGroupStore.getTabGroup(groupId)
  if (!groupId || groupId === -1 || !tabGroup) {
    return null
  }
  return (
    <hr
      className="absolute border"
      style={{
        width: 'calc(100% - 24px)',
        left: 12,
        bottom: 2,
        borderColor: tabGroup.color,
      }}
    />
  )
}

export default _TabGroupIndicator
