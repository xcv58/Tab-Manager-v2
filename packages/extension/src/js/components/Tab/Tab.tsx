import React, { useRef, useEffect } from 'react'
import { observer } from 'mobx-react-lite'
import Icon from 'components/Tab/Icon'
import TabTools from 'components/Tab/TabTools'
import TabContent from 'components/Tab/TabContent'
import CloseButton from 'components/CloseButton'
import classNames from 'classnames'
import { useStore } from 'components/hooks/useStore'
import { useTheme } from 'components/hooks/useTheme'
import { TabProps } from 'components/types'
import PIN from './Pin'
import ContainerOrGroupIndicator from './ContainerOrGroupIndicator'
import useReduceMotion from 'libs/useReduceMotion'

export default observer((props: TabProps & { className?: string }) => {
  const nodeRef = useRef(null)
  const { dragStore, searchStore } = useStore()
  const isDarkTheme = useTheme()
  const { tab, className } = props
  const {
    setNodeRef,
    isFocused,
    isMatched,
    isSelected,
    pinned,
    shouldHighlight,
  } = tab

  const isActionable = !dragStore.dragging

  const onMouseEnter = () => {
    if (isActionable) {
      tab.hover()
    }
  }

  const onMouseLeave = () => {
    if (isActionable) {
      tab.unhover()
    }
  }

  const onRemove = () => {
    const { removing, remove } = tab
    if (!removing) {
      remove()
    }
  }

  const reduceMotion = useReduceMotion()

  useEffect(() => {
    if (isFocused) {
      if (!searchStore.typing) {
        nodeRef.current.focus({ preventScroll: true })
        nodeRef.current.scrollIntoView({
          behavior: reduceMotion ? 'auto' : 'smooth',
        })
      }
    }
    return onMouseLeave
  }, [isFocused])
  useEffect(() => {
    setNodeRef(nodeRef)
  })

  const pin = pinned && PIN

  return (
    <div
      ref={nodeRef}
      tabIndex={-1}
      className={classNames(
        className,
        'flex relative items-center',
        {
          'opacity-25': !isMatched,
        },
        !className && [
          !isDarkTheme && {
            'hover:bg-blue-300': isActionable,
            'bg-blue-100': shouldHighlight,
            'bg-blue-300': isSelected,
          },
          isDarkTheme && {
            'hover:bg-gray-800': isActionable,
            'bg-gray-800': shouldHighlight,
            'bg-gray-900': isSelected,
          },
        ]
      )}
      onMouseEnter={onMouseEnter}
      onMouseOver={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {pin}
      <Icon tab={tab} />
      <TabContent tab={tab} />
      <TabTools tab={tab} />
      <CloseButton onClick={onRemove} disabled={tab.removing} />
      <ContainerOrGroupIndicator
        groupId={tab.groupId}
        cookieStoreId={tab.cookieStoreId}
      />
    </div>
  )
})
