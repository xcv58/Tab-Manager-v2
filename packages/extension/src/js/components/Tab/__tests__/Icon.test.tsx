import React from 'react'
import { shallow } from 'enzyme'
import Checkbox from '@mui/material/Checkbox'
import IconButton from '@mui/material/IconButton'
import { Icon } from 'components/Tab/Icon'

const tab = {
  focus: jest.fn(),
  select: jest.fn(),
  iconUrl: 'url',
  isSelected: false,
  bulkSelect: jest.fn(),
}

const props = { tab }

describe('Icon', () => {
  it('should render correct component', () => {
    const el = shallow(<Icon {...props} />)
    expect(el.find(IconButton).length).toBe(1)
    expect(el.find('img').length).toBe(1)
    expect(el.find(Checkbox).length).toBe(1)
  })

  it('should call correct function based on input', () => {
    const el = shallow(<Icon {...props} />)
    el.find(Checkbox).props().onClick({ shiftKey: false })
    expect(tab.select.mock.calls.length).toBe(1)
    expect(tab.bulkSelect.mock.calls.length).toBe(0)
    el.find(Checkbox).props().onClick({ shiftKey: true })
    expect(tab.select.mock.calls.length).toBe(1)
    expect(tab.bulkSelect.mock.calls.length).toBe(1)
    el.find(Checkbox).props().onClick({ shiftKey: true })
    expect(tab.select.mock.calls.length).toBe(1)
    expect(tab.bulkSelect.mock.calls.length).toBe(2)
  })

  it('should always call select if the isSelected is true', () => {
    const select = jest.fn()
    const bulkSelect = jest.fn()
    const el = shallow(
      <Icon tab={{ ...tab, select, bulkSelect, isSelected: true }} />
    )
    el.find(Checkbox).props().onClick({ shiftKey: false })
    expect(select.mock.calls.length).toBe(1)
    expect(bulkSelect.mock.calls.length).toBe(0)
    el.find(Checkbox).props().onClick({ shiftKey: true })
    expect(select.mock.calls.length).toBe(2)
    expect(bulkSelect.mock.calls.length).toBe(0)
    el.find(Checkbox).props().onClick({ shiftKey: true })
    expect(select.mock.calls.length).toBe(3)
    expect(bulkSelect.mock.calls.length).toBe(0)
  })
})
