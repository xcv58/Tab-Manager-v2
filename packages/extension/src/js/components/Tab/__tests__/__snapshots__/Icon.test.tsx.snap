// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should render correct component 1`] = `
<div>
  <div
    class="group"
  >
    <div
      class="group-hover:hidden"
    >
      <button
        aria-label="Toggle select"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium focus:outline-none focus:ring css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <img
          class="w-6 h-6"
          src="url"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="focus:outline-none focus:ring hidden group-hover:block"
    >
      <span
        class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="Toggle select"
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
    </div>
  </div>
</div>
`;
